indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/PerfectRollupWorkerTask.java

```diff
@@ -0,0 +1,129 @@
+ * under the License.
+ */
+
+package org.apache.druid.indexing.common.task.batch.parallel;
+
+import com.google.common.base.Preconditions;
+import org.apache.druid.indexer.partitions.PartitionsSpec;
+import org.apache.druid.indexing.common.actions.TaskActionClient;
+import org.apache.druid.indexing.common.task.AbstractBatchIndexTask;
+import org.apache.druid.indexing.common.task.TaskResource;
+import org.apache.druid.indexing.common.task.Tasks;
+import org.apache.druid.java.util.common.granularity.Granularity;
+import org.apache.druid.segment.indexing.DataSchema;
+import org.apache.druid.segment.indexing.granularity.ArbitraryGranularitySpec;
+import org.apache.druid.segment.indexing.granularity.GranularitySpec;
+import org.apache.druid.timeline.DataSegment;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Base class for parallel indexing perfect rollup worker tasks.
+ */
+abstract class PerfectRollupWorkerTask extends AbstractBatchIndexTask
+{
+  private final GranularitySpec granularitySpec;
+  private final DataSchema dataSchema;
+  private final ParallelIndexTuningConfig tuningConfig;
+
+  PerfectRollupWorkerTask(
+      String id,
+      @Nullable String groupId,
+      @Nullable TaskResource taskResource,
+      DataSchema dataSchema,
+      ParallelIndexTuningConfig tuningConfig,
+      @Nullable Map<String, Object> context
+  )
+  {
+    super(id, groupId, taskResource, dataSchema.getDataSource(), context);
+
+    Preconditions.checkArgument(
+        tuningConfig.isForceGuaranteedRollup(),
+        "forceGuaranteedRollup must be set"
+    );
+
+    checkPartitionsSpec(tuningConfig.getGivenOrDefaultPartitionsSpec());
+
+    granularitySpec = dataSchema.getGranularitySpec();
+    Preconditions.checkArgument(
+        !granularitySpec.inputIntervals().isEmpty(),
+        "Missing intervals in granularitySpec"
+    );
+
+    this.dataSchema = dataSchema;
+    this.tuningConfig = tuningConfig;
+  }
+
+  private static void checkPartitionsSpec(PartitionsSpec partitionsSpec)
+  {
+    if (!partitionsSpec.isForceGuaranteedRollupCompatible()) {
+      String incompatibiltyMsg = partitionsSpec.getForceGuaranteedRollupIncompatiblityReason();
+      String msg = "forceGuaranteedRollup is incompatible with partitionsSpec: " + incompatibiltyMsg;
+      throw new IllegalArgumentException(msg);
+    }
+  }
+
+  @Override
+  public final int getPriority()
+  {
+    return getContextValue(Tasks.PRIORITY_KEY, Tasks.DEFAULT_BATCH_INDEX_TASK_PRIORITY);
+  }
+
+  @Override
+  public final boolean requireLockExistingSegments()
+  {
+    return true;
+  }
+
+  @Override
+  public final List<DataSegment> findSegmentsToLock(TaskActionClient taskActionClient, List<Interval> intervals)
+  {
+    throw new UnsupportedOperationException("This task locks by timeChunk instead of segment");
+  }
+
+  @Override
+  public final boolean isPerfectRollup()
```

Comment:
nit: It looks like the intention of this is to indicate that we want to use the timeChunk for rollups instead of segment?  Should we rename it to something like `useTimeChunkForRollups` It's unclear to me why perfect Rollups mean using the timeChunk instead of segment for rollups.

Also it looks like this is only used in `AbstractBatchIndexTask` can we make this method protected instead?

Reply:
I'm not sure what you mean by "using timeChunk for rollups", but the perfect rollup is pre-aggregating rows across the entire input data. See https://druid.apache.org/docs/latest/ingestion/index.html#rollup for more details.

Reply:
@jihoonson In `AbstractBatchIndexTask` line 268-271 it says 
```
if (isPerfectRollup()) {
    log.info("Using timeChunk lock for perfect rollup");
    ...
}
```

Reply:
Ah, it's the `timeChunk lock` rather than `timeChunk`. Tasks must use the timeChunk lock instead of segment lock for perfect rollup because the rollup happens across the entire input data. See https://github.com/apache/incubator-druid/issues/7491 for timeChunk lock vs segment lock.